### PR TITLE
[bridges-v2] Permissionless lanes 

### PR DESCRIPTION
**Pull request by bkontur**
Relates to: https://github.com/paritytech/parity-bridges-common/issues/2451
Closes: https://github.com/paritytech/parity-bridges-common/issues/2500

## Summary

Now, the bridging pallet supports only static lanes, which means lanes that are hard-coded in the runtime files. This PR fixes that and adds support for dynamic, also known as permissionless, lanes. This means that allowed origins (relay chain, sibling parachains) can open and close bridges (through BridgeHubs) with another bridged (substrate-like) consensus using just `xcm::Transact` and `OriginKind::Xcm`.

_This PR is based on the migrated code from the Bridges V2 [branch](https://github.com/paritytech/polkadot-sdk/pull/4427) from the old `parity-bridges-common` [repo](https://github.com/paritytech/parity-bridges-common/tree/bridges-v2)._

## Explanation

Please read [bridges/modules/xcm-bridge-hub/src/lib.rs](https://github.com/paritytech/polkadot-sdk/blob/149b0ac2ce43fba197988f2642032fa24dd8289a/bridges/modules/xcm-bridge-hub/src/lib.rs#L17-L136) to understand how managing bridges works. The basic concepts around `BridgeId` and `LaneId` are also explained there.

## TODO

- [x] search and fix for comment: `// TODO:(bridges-v2) - most of that stuff was introduced with free header execution: https://github.com/paritytech/polkadot-sdk/pull/4102` - more info in the comment [bellow](https://github.com/paritytech/polkadot-sdk/pull/4427#issuecomment-2126625043)
- [x] TODO: there's only one impl of `EnsureOrigin<Success = Location>`

## TODO - not blocking review

**benchmarking:**
- [x] regenerate all relevant weights for BH/AH runtimes
- [ ] regenerate default weights for bridging pallets e.g. `modules/messages/src/weights.rs`
- [ ] add benchmarks for `xcm-bridge-hub` pallet https://github.com/paritytech/polkadot-sdk/issues/5550

**testing:**
- [ ] add xcm-emulator tests for Rococo/Penpal to Westend/Penpal with full opening channel and sending/receiving `xcm::Transact`

**migrations:**
- [x] add migrations for BridgeHubRococo/Westend https://github.com/paritytech/parity-bridges-common/issues/2794 (to be reusable for P/K bridge)
  - [x] check also storage migration, if needed for pallets
  - [ ] migration for XCM type (optional)
  - [x] migration for static lanes to the dynamic (reuse for fellows)

**investigation:**
- [ ] revisit https://github.com/paritytech/parity-bridges-common/issues/2380
- [ ] check congestion around `LocalXcmChannelManager` and `OutboundLanesCongestedSignals` impls - https://github.com/paritytech/polkadot-sdk/issues/5551
  - to be reusable for polkadot-fellows
  - return `report_bridge_status` was remove, so we need to `XcmpQueue` alternative?

### CODE REVIEW

**Comment by bkontur** [comment]
bot help

File: bridges/modules/messages/src/lib.rs
Comment (franciscoaguirre) [inline bridges/modules/messages/src/lib.rs:278]: I thought messages could be delivered out-of-order. This means if there's a really big message it'll clog the lane until some relayer is willing to send it?

**Review by franciscoaguirre — COMMENTED** [review]
I did one review round, will do more later

File: bridges/modules/messages/src/lanes_manager.rs
Comment (bkontur) [reply bridges/modules/messages/src/lanes_manager.rs]: thx, you're right, removed here: https://github.com/paritytech/polkadot-sdk/pull/4949/commits/ad93c2b0c8114a0e014d04ef532c46f46fa9bca1

File: bridges/bin/runtime-common/src/extensions/mod.rs
Comment (bkontur) [reply bridges/bin/runtime-common/src/extensions/mod.rs:21]: They were refactored and divided into several modules

File: bridges/modules/messages/src/lib.rs
Comment (bkontur) [reply bridges/modules/messages/src/lib.rs:278]: I don't remember exact reasoning, but the order is crucial for one lane, e.g. there is a description:
> Messages that are sent over the same lane are guaranteed to be delivered to the target chain in
the same order they're sent from the source chain. In other words, message with nonce `N` will be delivered right before
delivering a message with nonce `N+1`.

I think it is about some guarantees, I think that HRMP/UMP/DMP works also  in the order.
I think when you process message between two points in random order it can allow attacks, e.g. if you send two messages (in order: msg1 and then msg2):
```
msg1: AHP -> deposit(a1) -> AHK
msg2: AHP -> do_whatever_with(a1) -> AHK
```
we cannot just randomly change the order.

**Comment by bkontur** [comment]
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_grandpa
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_parachains
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_messages
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --pallet=pallet_bridge_relayers
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_grandpa
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_parachains
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_messages
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --pallet=pallet_bridge_relayers

**Comment by bkontur** [comment]
bot bench cumulus-bridge-hubs --runtime=bridge-hub-rococo --subcommand=xcm --pallet=pallet_xcm_benchmarks::generic
bot bench cumulus-bridge-hubs --runtime=bridge-hub-westend --subcommand=xcm --pallet=pallet_xcm_benchmarks::generic
bot bench cumulus-assets --runtime=asset-hub-rococo --pallet=pallet_xcm_bridge_hub_router
bot bench cumulus-assets --runtime=asset-hub-westend --pallet=pallet_xcm_bridge_hub_router
bot bench cumulus-assets --runtime=asset-hub-rococo --subcommand=xcm --pallet=pallet_xcm_benchmarks::fungible
bot bench cumulus-assets --runtime=asset-hub-rococo --subcommand=xcm --pallet=pallet_xcm_benchmarks::generic
bot bench cumulus-assets --runtime=asset-hub-westend --subcommand=xcm --pallet=pallet_xcm_benchmarks::fungible
bot bench cumulus-assets --runtime=asset-hub-westend --subcommand=xcm --pallet=pallet_xcm_benchmarks::generic

File: bridges/modules/xcm-bridge-hub/src/exporter.rs
Comment (franciscoaguirre) [inline bridges/modules/xcm-bridge-hub/src/exporter.rs:217]:
```suggestion
		// messages to the suspended bridge, hoping that it'll be actually resumed soon
```

?

File: bridges/modules/xcm-bridge-hub/src/exporter.rs
Comment (franciscoaguirre) [inline bridges/modules/xcm-bridge-hub/src/exporter.rs:270]: Should this be `OUTBOUND_LANE_CONGESTED_THRESHOLD` instead? Since the variable is called `is_congested`

Comment (franciscoaguirre) [reply bridges/modules/xcm-bridge-hub/src/exporter.rs:270]: Do `enqueued_messages` decrease when the message is delivered? Giving this a chance of being resumed

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (franciscoaguirre) [inline bridges/modules/xcm-bridge-hub/src/lib.rs:289]: I thought the bridge was between two global consensuses and then different parachains in these consensuses had lanes between them. But given these functions, these connections would be bridges? What's the model to think about lanes then?

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (franciscoaguirre) [inline bridges/modules/xcm-bridge-hub/src/lib.rs:228]: I'd call it `Fungible` to not confuse it with the old Currency traits, which we still have around 😅

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (franciscoaguirre) [inline bridges/modules/xcm-bridge-hub/src/lib.rs:710]: We shouldn't add new `getter`s since [they are deprecated](https://github.com/paritytech/polkadot-sdk/issues/3326)

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (franciscoaguirre) [inline bridges/modules/xcm-bridge-hub/src/lib.rs:719]:
```suggestion
		/// Keep in mind that we are **NOT** reserving any amount for the bridges opened at
```

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (franciscoaguirre) [inline bridges/modules/xcm-bridge-hub/src/lib.rs:724]:
```suggestion
	_phantom: sp_std::marker::PhantomData<(T, I)>,
```

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (bkontur) [reply bridges/modules/xcm-bridge-hub/src/lib.rs]: fixed here: https://github.com/paritytech/polkadot-sdk/pull/4949/commits/ea9adbb3d3d717fbb38c66a8bfb675a64bf1e399

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (bkontur) [reply bridges/modules/xcm-bridge-hub/src/lib.rs:228]: well, I used `Currency`, because it is used everywhere. I only see one `type Fungibles:`, but no `type Fungible:`. So, I think it would be better to rename this in all other places as well, but as a separate PR.

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (bkontur) [reply bridges/modules/xcm-bridge-hub/src/lib.rs]: @franciscoaguirre I had to revert this: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/7048190

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (bkontur) [reply bridges/modules/xcm-bridge-hub/src/lib.rs:294]: @franciscoaguirre 
BridgeHubs on both side manages lanes (basically, queues or structs which holds some metadata) and the lane can be created e.g. between two parachains from different consensus, the lane can be also between relaychains.

E.g.

Lane_01:  AssetHubPolkadot -> BridgeHubPolkadot <> BridgeHubKusama <- AssetHubKusama
Lane_02:  Acala -> BridgeHubPolkadot <> BridgeHubKusama <- Encointer
Lane_03:  Polkadot -> BridgeHubPolkadot <> BridgeHubKusama <- Encointer
Lane_04:  CollectivesPolkadot -> BridgeHubPolkadot <> BridgeHubKusama <- Kusama

Please, check the example in the docs: https://github.com/paritytech/polkadot-sdk/blob/b974795ba8984471cb76b254ee41e485c8d0ce0e/bridges/modules/xcm-bridge-hub/src/lib.rs#L114-L141
If it is still not clear or missing anything, let's fix the docs here.

**Comment by bkontur** [comment]
bot clean

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (franciscoaguirre) [reply bridges/modules/xcm-bridge-hub/src/lib.rs:294]: They are still called "bridges" as well as "lanes" and are opened with `fn open_bridge` instead of something like `fn open_lane`? That's what's a little confusing, nothing more

File: cumulus/parachains/runtimes/bridge-hubs/common/Cargo.toml
Comment (franciscoaguirre) [inline cumulus/parachains/runtimes/bridge-hubs/common/Cargo.toml:15]: We shouldn't keep on using `sp-std`: https://github.com/paritytech/polkadot-sdk/issues/2101

File: bridges/modules/xcm-bridge-hub/src/lib.rs
Comment (bkontur) [reply bridges/modules/xcm-bridge-hub/src/lib.rs:294]: I would say that, from the caller's perspective, it is a "bridge to another consensus somewhere." The pallet_xcm_bridge_hub stores Bridges (bridge metadata, such as who is connected with whom, deposits, and any other details needed to manage the bridge), while the lane is a hidden implementation detail. The lanes are stored in the pallet_bridge_messages pallet, and the caller doesn't need to know about the lanes or other specifics—only that someone needs to run a relayer for this lane.

**Review by franciscoaguirre — APPROVED** [review]

**Review by serban300 — APPROVED** [review]
Looks good ! Just left a few nits.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-kusama-bridge/2971/61

**Comment by hqwangningbo** [comment]
@bkontur @serban300 Hey guys, can you speed up that feature?🙏
Bifrost wants to implement cross-chaining of BNC between Polkadot-Kusama. Currently this feature is implemented through centralised exchanges, but it's not user friendly.
We need to implement BNC across the chain using TeleportAssets.

**Comment by serban300** [comment]
> @bkontur @serban300 Hey guys, can you speed up that feature?🙏 Bifrost wants to implement cross-chaining of BNC between Polkadot-Kusama. Currently this feature is implemented through centralised exchanges, but it's not user friendly. We need to implement BNC across the chain using TeleportAssets.

@hqwangningbo for me, apart from a couple of nits, this PR is good to go, but we still need a security audit for it and after that there are still a couple of things that are left to be done as I understand from the description. Also @bkontur is on vacation this week.

Also after the development is complete, I think we'll still need to wait until the next release window (which I think is once every 3 months ?) to deploy this on Polkadot. But I'm not 100% sure. Maybe @acatangiu or @bkontur have more details about this feature and the timeline.

@hqwangningbo when would you need this ? Do you have a specific date in mind ?

**Comment by hqwangningbo** [comment]
> > @bkontur @serban300 Hey guys, can you speed up that feature?🙏 Bifrost wants to implement cross-chaining of BNC between Polkadot-Kusama. Currently this feature is implemented through centralised exchanges, but it's not user friendly. We need to implement BNC across the chain using TeleportAssets.
> 
> @hqwangningbo for me, apart from a couple of nits, this PR is good to go, but we still need a security audit for it and after that there are still a couple of things that are left to be done as I understand from the description. Also @bkontur is on vacation this week.
> 
> Also after the development is complete, I think we'll still need to wait until the next release window (which I think is once every 3 months ?) to deploy this on Polkadot. But I'm not 100% sure. Maybe @acatangiu or @bkontur have more details about this feature and the timeline.
> 
> @hqwangningbo when would you need this ? Do you have a specific date in mind ?

For me of course the sooner the better, but I understand you have a lot of work to do. It would be helpful to have a clear timeline for us as well.

**Comment by lurpis** [comment]
It has been more than a year and a half since I first heard about the P <> K bridge. Is there any possibility of speeding it up? Since Bifrost assets are distributed on two parachains, Bifrost-Kusama and Bifrost-Polkadot, we need to rely on the opening of the bridge before launching the tokenomics 2.0. but Bifrost Tokenomics 2.0 is scheduled to be released in October.

**Comment by acatangiu** [comment]
@lurpis 

> It has been more than a year and a half since I first heard about the P <> K bridge. Is there any possibility of speeding it up? Since Bifrost assets are distributed on two parachains, Bifrost-Kusama and Bifrost-Polkadot, we need to rely on the opening of the bridge before launching the tokenomics 2.0. but Bifrost Tokenomics 2.0 is scheduled to be released in October.

The P <> K bridge is live since April this year. It supports bi-directional asset transfers since then.

IIUC correctly, what you want is custom bridged token handling: have dedicated bridge lane between Bifrost-Kusama and Bifrost-Polkadot so you can directly transfer PolkadotBNC and KusamaBNC between the two (without going through Asset Hubs), and have custom logic on your parachains to treat the two assets as one single/global BNC asset.

@hqwangningbo 

> For me of course the sooner the better, but I understand you have a lot of work to do. It would be helpful to have a clear timeline for us as well.

The *permissionless lanes* feature (this PR) is currently under security audit, but we are confident it will be released with the end-of-September Polkadot-SDK release. The new SDK components will then have to be picked up by https://github.com/polkadot-fellows/runtimes/ maintainers and released with a new Runtimes release, then those new BridgeHub runtimes needs to be upgraded on-chain.

An optimistic estimate for all of the above is Nov 2024.

> Bifrost Tokenomics 2.0 is scheduled to be released in October.

You could develop this in parallel, test it out on Zombienet or local networks (or Rococo Westend if you have Bifrost chains deployed there), and when BH is updated ~Nov, you can be ready and launch immediately.

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs
Comment (acatangiu) [inline cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs:141]: This should be only `People` System Chain. We don't allow (yet) other (especially non-system) parachains to open bridge to Bulletin.

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs
Comment (acatangiu) [inline cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs:151]: same filter here - only people-chain

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs
Comment (bkontur) [reply cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs:141]: good point, then we can even set it to `()` and disable it, 
because there is also another feature, where we can open channel with migration and for the bulletin it is already set it up: https://github.com/paritytech/polkadot-sdk/pull/4949/files#diff-e64ba9883926418979cf0f5886d108a30a96518d345b2ba819bcde75cbd8aa98R291-R298
```
	/// Ensure that the existing lanes for the People<>Bulletin bridge are correctly configured.
	pub type StaticToDynamicLanes = pallet_xcm_bridge_hub::migration::OpenBridgeForLane<
		Runtime,
		XcmOverPolkadotBulletinInstance,
		RococoPeopleToRococoBulletinMessagesLane,
		ConstBool<true>,
		PeopleRococoLocation,
		BulletinRococoLocation,
	>;
```

Comment (bkontur) [reply cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs:141]: actually, `OpenBridgeOrigin = ()` is not good, because People can also decide to close channel, I will set it for just People

Comment (bkontur) [reply cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs]: fixed here: https://github.com/paritytech/polkadot-sdk/pull/4949/commits/513aa6aeecdb7ebd7e8531cdca76ad8b36781f33

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs
Comment (bkontur) [reply cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/bridge_to_bulletin_config.rs]: fixed here: https://github.com/paritytech/polkadot-sdk/pull/4949/commits/513aa6aeecdb7ebd7e8531cdca76ad8b36781f33

File: bridges/modules/messages/README.md
Comment (acatangiu) [inline bridges/modules/messages/README.md:34]: You should add updated documentation here - how are lanes opened and used.

File: bridges/modules/messages/src/lanes_manager.rs
Comment (acatangiu) [reply bridges/modules/messages/src/lanes_manager.rs:187]: +1

**Comment by acatangiu** [comment]
FYI: @lurpis @hqwangningbo there's a public [Polkadot-SDK Release Calendar](https://github.com/orgs/paritytech/projects/161) you can now use. The individual releases boards (e.g. [Stable2409 LTS Release](https://github.com/orgs/paritytech/projects/162)) will also be populated with the PRs part of said releases.

**Comment by lurpis** [comment]
@acatangiu 
> IIUC correctly, what you want is custom bridged token handling: have dedicated bridge lane between Bifrost-Kusama and Bifrost-Polkadot so you can directly transfer PolkadotBNC and KusamaBNC between the two (without going through Asset Hubs), and have custom logic on your parachains to treat the two assets as one single/global BNC asset.

Yes, we completed the logic here in August and passed on the testnet. What we are waiting for now is the restriction to be lifted.

**Comment by acatangiu** [comment]
> Yes, we completed the logic here in August and passed on the testnet. What we are waiting for now is the restriction to be lifted.

@lurpis @hqwangningbo can you set up a Telegram group chat or Matrix group chat to discuss your plan and needs in detail?

I want to make sure we're not missing anything and not delaying things more.

my Telegram: @adriankata
my Matrix: @adrian:parity.io

**Comment by hqwangningbo** [comment]
@acatangiu I have invited you to the Parity-Bifrost Telegram group and hope to have further communication.

**Pull request merged**
